### ISSUE TITLE
libpmemobj: function pointer is cast to an incompatible function pointer

### ISSUE THREAD

**Issue by mbiesek** [pmem/issues#867]
In place mentioned below:

https://github.com/pmem/pmdk/blob/3b5b03fe00749400e20ab9d8f180cb1728709776/src/libpmemobj/obj.c#L1033-L1034

There is function pointer cast to an incompatible function pointer.

According to standard:
Section 6.3.2.3/8 "Pointers" C99
"A pointer to a function of one type may be converted to a pointer to a function of another type and back again; the result shall compare equal to the original pointer. **If a converted pointer is used to call a function whose type is not compatible with the pointed-to type, the behavior is undefined.**"

**Intention of this issue is to point potential ill-defined place** ( Undefined Behaviour).

Unfortunately I could not produce crash or "nasal demons" in source code, but  In case of UB it strictly depend from compiler implementation.

Only sanitizer which I know points that this is UB is in clang:
-fsanitize=function but it supports only C++ 

I created example code - based on libpmembj code mentioned above.
I used C++ compiler, which supports diagnostics, in code below in output, You could see runtime error

https://wandbox.org/nojs/clang-6.0.0/permlink/R1y3fdPcpKdkFQmx

**Comment by mbiesek** [comment]
I think You could consider in fixing this issue, to add in Makefiles warning flag **-Wcast-function-type** , which could prevent to repeat this situation in future.
Be aware that this flag was is present in GCC 8.1 (Released: May 2, 2018) 

For support this diagnostic in CI process maybe it is also good idea to support recently released OS, like: **Ubuntu 18.04 LTS Bionic Beaver**  (Released April 26, 2018 ) and **Fedora 28** ( Released May 1 2018)   which support this version of GCC - it's just a idea which I think could be realized as separate issue.

**Comment by marcinslusarz** [comment]
PMDK has a mechanism for detecting whether flag is supported or not (see usage of check_flag function in src/Makefile.inc), but currently I don't have access to any system with this gcc version, so I can't verify PMDK builds with this flag. Feel free to test it and provide a patch once #2933 will land.

If we'll enable this flag we also have to regularly build PMDK on system with that flag. We can't do this on Travis (too many jobs already and updating to the latest version is a no-go), but we can do this internally on Jenkins. I'll ask some folks to set it up once  -Wcast-function-type patch will land.

**Comment by lukaszstolarczuk** [comment]
merged and backported to 1.4, should works fine.
